src/chains/ethereum/ethereum/src/helpers/simulation-handler.ts

```diff
@@ -0,0 +1,477 @@
+    } else if (this.#runCallOpts) {
+      callResult = await this.#vm.runCall(this.#runCallOpts);
+      this.#emitAfter();
+    } else {
+      throw new Error("Unknown error when simulating transaction.");
+    }
+
+    if (callResult.execResult.exceptionError) {
+      throw new CallError(callResult);
+    } else {
+      return callResult;
+    }
+  }
+
+  /**
+   * Recursively runs the `vm.runCall` function until the touched addresses and
+   * storage are the same as the previously generated access list.
+   * @param initialAccessList An access list to initially run the transaction with.
+   * @returns The final access list generated, plus an estimate of the gas
+   * consumed by the running the transaction _with_ the generated access list included.
+   */
+  public async createAccessList(initialAccessList: AccessList) {
+    await this.#stateManager.checkpoint();
+    return await this.#createAccessList(initialAccessList);
+  }
+
+  async #createAccessList(
+    previousAccessList: AccessList
+  ): Promise<{ accessList: AccessList; gasUsed: string }> {
+    const stateManager = this.#stateManager;
+    const callResult = await this.runCall();
+    const accessList = stateManager.generateAccessList(
+      this.#accessListExclusions,
+      this.#addressesOnlyStorage
+    );
+    if (JSON.stringify(previousAccessList) === JSON.stringify(accessList)) {
+      const { dataFeeEIP2930 } = AccessLists.getAccessListData(accessList);
+      const baseFeeBN = new BN(
+        Quantity.toBuffer(this.#intrinsicGas + dataFeeEIP2930)
+      );
+      const gasUsedBN = callResult.gasUsed.add(baseFeeBN);
+      const gasUsed = Quantity.toString(gasUsedBN.toBuffer());
+      return { accessList, gasUsed };
+    } else {
+      await stateManager.revert();
+      return await this.#createAccessList(accessList);
+    }
+  }
+
+  /**
+   * Calls the StateManager's `addWarmedAddress` function for the transaction
+   * caller, all precompiles, and the to address, if applicable.
+   * @param stateManager
+   * @param caller
+   * @param to
+   * @returns An array of precompile addresses
+   */
+  #warmDefaults(caller: EthereumJsAddress, to?: EthereumJsAddress) {
+    const stateManager = this.#stateManager;
+    let precompiles: EthereumJsAddress[] = [];
+    // handle Berlin hardfork warm storage reads
+    precompiles = warmPrecompiles(stateManager);
+    stateManager.addWarmedAddress(caller.buf);
+    if (to) stateManager.addWarmedAddress(to.buf);
+
+    return precompiles;
+  }
+
+  #warmAccessList(accessList: AccessList) {
+    const stateManager = this.#stateManager;
+    for (const { address, storageKeys } of accessList) {
+      const addressBuf = Address.from(address).toBuffer();
+      stateManager.addWarmedAddress(addressBuf);
+      for (const slot of storageKeys) {
+        stateManager.addWarmedStorage(addressBuf, Data.toBuffer(slot, 32));
+      }
+    }
+  }
+
+  async #applySimulationOverrides(overrides: CallOverrides): Promise<void> {
```

Comment:
This function is largely unchanged from how it was in the run-call file. To port it over to here, the function no longer receives `stateTrie` and `vm` parameters and just pulls it from the class properties instead.